phase/src/caller/caller_algorithm.cpp

```diff
@@ -97,45 +115,85 @@ void caller::phase_iteration() {
 		vrb.progress("  * HMM imputation", prog_bar);
 	}
 	vrb.bullet("HMM imputation [#states=" + stb.str(statH.mean(), 1) + " / %poly=" + stb.str(statC.mean(), 1) + "%] (" + stb.str(tac.rel_time()*1.0/1000, 2) + "s)");
-}
 
-void caller::phase_loop() {
-	//First Iteration
-	current_stage = STAGE_INIT;
-	vrb.title("Initializing iteration");
 
-	H.initRareTar(G,V);
-	H.performSelection_RARE_INIT_GL(V); //not parallel
+	write_checkpoint();
 
-	phase_iteration();
+	if (current_stage == STAGE_INIT) {
+		H.init_states.clear();
+		H.init_states.shrink_to_fit();
+	}
+}
 
-	H.init_states.clear();
-	H.init_states.shrink_to_fit();
+void caller::increment_iteration() {
+	current_iteration++;
+	if (current_iteration == iterations_per_stage[current_stage]) {
+		current_stage++;
+		current_iteration = 0;
+	}
+}
 
-	//Burn-in 0
-	current_stage = STAGE_BURN;
-	int nBurnin = options["burnin"].as < int > ();
-	for (int iter = 0 ; iter < nBurnin ; iter ++)
-	{
-		vrb.title("Burn-in iteration [" + stb.str(iter+1) + "/" + stb.str(nBurnin) + "]");
-		H.updateHaplotypes(G);
-		H.transposeRareTar();
-		H.matchHapsFromCompressedPBWTSmall(V, false);
-		//current_stage = STAGE_RESTRICT;
-		//phase_iteration();
-		current_stage = STAGE_BURN;
-		phase_iteration();
+void caller::write_checkpoint() {
+	if (options.count("checkpoint-file-out")) {
+		
+		vrb.bullet("writing out checkpoint");
+		std::string cp_filename = options["checkpoint-file-out"].as < std::string > ();
+		std::string tmp_cp_filename = cp_filename + ".tmp";
+		std::ofstream ofs(tmp_cp_filename, std::ios::binary | std::ios_base::out);
+		boost::archive::binary_oarchive oa(ofs);
+		tac.clock();
+		oa << current_stage;
+		oa << current_iteration;
+		oa << G.vecG.size();  //n_samples
+		std::cout<<"G.vecG.size()=="<<G.vecG.size()<<std::endl;
+		std::cout<<"current_stage="<<current_stage<<std::endl;
+		std::cout<<"current_iteration="<<current_iteration<<std::endl;
+		for (size_t i =0; i < G.vecG.size(); i++)
+		{
+			oa << G.vecG[i]->stored_cnt;
+			oa << G.vecG[i]->stored_data;
+
+			oa << G.vecG[i]->H0;
+			oa << G.vecG[i]->H1;
+		}
+		std::rename(tmp_cp_filename.c_str(), cp_filename.c_str());
```

Comment:
this was my attempt to guarantee that the checkpoint file is never in an incomplete state when another process can access it (to copy it into cloud storage, for example).  as far as I can tell, not all unix systems provide mandatory system wide locks, so this was the best I could come up with.  I'm not completely sure that this guarantees the objective though, and if there's a better way to do this would be great to know.

Reply:
switched this to `std::filesystem::rename` based on https://en.cppreference.com/w/cpp/io/c/rename which states that 

> If new_filename exists, the behavior is implementation-defined.

>  [POSIX](https://pubs.opengroup.org/onlinepubs/9699919799/functions/rename.html) specifies many additional details on the semantics of this function, which are reproduced in C++ by [std::filesystem::rename](https://en.cppreference.com/w/cpp/filesystem/rename).

also checked w/ some other folks who confirmed this is a commonly used technique for achieving this behavior.